File: offchain/core/src/strategy/gc.rs
use std::{collections::HashMap, error::Error};

use ::log::info;
use async_recursion::async_recursion;
use ethers::types::Address;

use crate::arena::{Arena, MatchState, TournamentState};

#[derive(Debug)]
pub enum PlayerTournamentResult {
    TournamentWon,
    TournamentLost,
}

pub struct GarbageCollector<A: Arena> {
    arena: A,
    root_tournamet: Address,
}

impl<A: Arena> GarbageCollector<A> {
    pub fn new(arena: A, root_tournamet: Address) -> Self {
        Self {
            arena,
            root_tournamet,
        }
    }

    pub async fn react(&mut self) -> Result<(), Box<dyn Error>> {
        let tournament_states = self.arena.fetch_from_root(self.root_tournamet).await?;
        self.react_tournament(self.root_tournamet, tournament_states)
            .await
    }

    #[async_recursion]
    async fn react_tournament(
        &mut self,
        tournament_address: Address,
        tournament_states: HashMap<Address, TournamentState>,
    ) -> Result<(), Box<dyn Error>> {
        info!("Enter tournament at address: {}", tournament_address);
        let tournament_state = tournament_states
            .get(&tournament_address)
            .expect("tournament state not found");

        for m in tournament_state.matches.clone() {
            self.react_match(&m, tournament_states.clone()).await?;

            let status_1 = tournament_state
                .commitment_states
                .get(&m.id.commitment_one)
                .expect("status of commitment 1 not found");
            let status_2 = tournament_state
                .commitment_states
                .get(&m.id.commitment_two)
                .expect("status of commitment 2 not found");
            if (!status_1.clock.has_time()
                && (status_1.clock.time_since_timeout() > status_2.clock.allowance))
                || (!status_2.clock.has_time()
                    && (status_2.clock.time_since_timeout() > status_1.clock.allowance))
            {
                info!(
                    "eliminate match for commitment {} and {} at tournament {} of level {}",
                    m.id.commitment_one,
                    m.id.commitment_two,
                    tournament_address,
                    tournament_state.level
                );

                self.arena
                    .eliminate_match(tournament_address, m.id)
                    .await
                    .expect("fail to eliminate match");
            }
        }
        Ok(())
    }

    #[async_recursion]
    async fn react_match(
        &mut self,
        match_state: &MatchState,
        tournament_states: HashMap<Address, TournamentState>,
    ) -> Result<(), Box<dyn Error>> {
        info!("Enter match at HEIGHT: {}", match_state.current_height);
        if let Some(inner_tournament) = match_state.inner_tournament {
            return self
                .react_tournament(inner_tournament, tournament_states)
                .await;
        }

        Ok(())
    }
}

File: prt/client-lua/player/gc.lua
local helper = require "utils.helper"

local GarbageCollector = {}
GarbageCollector.__index = GarbageCollector

function GarbageCollector:new(sender)
    local gc_strategy = {
        sender = sender
    }

    setmetatable(gc_strategy, self)
    return gc_strategy
end

function GarbageCollector:_react_match(match)
    helper.log_full(self.sender.index,
        string.format("Garbage collect match at HEIGHT: %d, of tournament: %s", match.current_height,
            match.tournament.address))
    if match.inner_tournament then
        return self:_react_tournament(match.inner_tournament)
    end
end

function GarbageCollector:_react_tournament(tournament)
    for _, match in ipairs(tournament.matches) do
        if match then
            local status_1 = tournament.commitments[match.commitment_one].status
            local status_2 = tournament.commitments[match.commitment_two].status

            self:_react_match(match)

            -- try to eliminate matches that both clocks are out of time
            if (not status_1.clock:has_time() and
                    (status_1.clock:time_since_timeout() > status_2.clock.allowance)) or
                (not status_2.clock:has_time() and
                    (status_2.clock:time_since_timeout() > status_1.clock.allowance)) then
                helper.log_full(self.sender.index, string.format(
                    "eliminate match for commitment %s and %s at tournament %s of level %d",
                    match.commitment_one,
                    match.commitment_two,
                    tournament.address,
                    tournament.level
                ))

                local ok, e = self.sender:eliminate_match(
                    tournament.address,
                    match.commitment_one,
                    match.commitment_two
                )
                if not ok then
                    helper.log_full(self.sender.index, string.format(
                        "eliminate match reverted: %s",
                        e
                    ))
                end
            end
        end
    end
end

function GarbageCollector:react(root_tournament)
    return self:_react_tournament(root_tournament)
end

return GarbageCollector
